orttraining/orttraining/core/framework/module_gradient_graph_builder.cc

```diff
@@ -0,0 +1,393 @@
+    if (input_names.find(initializer.first) == input_names.end()) {
+      initializer_names_to_remove.insert(initializer.first);
+    }
+  }
+
+  for (const auto& initializer_name : initializer_names_to_remove) {
+    graph.RemoveInitializedTensor(initializer_name);
+  }
+}
+
+Status ModuleGradientGraphBuilder::Initialize(std::istream& model_istream,
+                                              const ModuleGradientGraphBuilderConfiguration& config) {
+  // Save the model and config.
+  ONNX_NAMESPACE::ModelProto model_proto;
+  ORT_RETURN_IF_ERROR(Model::Load(model_istream, &model_proto));
+  ORT_RETURN_IF_ERROR(Model::Load(model_proto, model_, nullptr, *logger_));
+  config_ = config;
+
+  // Handle original model inputs, outputs and trainable initializers.
+  // We need to move the trainable initializers to graph inputs and keep the order in config,
+  // it's possible that the graph already has some trainable initializers in graph inputs,
+  // so we need to NOT assign these trainable initializers to the user inputs list.
+  Graph& graph = model_->MainGraph();
+  std::unordered_set<std::string> initializer_names_to_train_set(config.initializer_names_to_train.begin(),
+                                                                 config.initializer_names_to_train.end());
+  const std::vector<const NodeArg*>& graph_inputs = graph.GetInputsIncludingInitializers();
+  for (auto& node_arg : graph_inputs) {
+    if (initializer_names_to_train_set.find(node_arg->Name()) == initializer_names_to_train_set.end()) {
+      split_graphs_info_.user_input_names.emplace_back(node_arg->Name());
+    }
+  }
+
+  const std::vector<const NodeArg*>& graph_outputs = graph.GetOutputs();
+  for (auto& node_arg : graph_outputs) {
+    split_graphs_info_.user_output_names.emplace_back(node_arg->Name());
+  }
+
+  split_graphs_info_.initializer_names_to_train.assign(config.initializer_names_to_train.begin(),
+                                                       config.initializer_names_to_train.end());
+
+  std::vector<const NodeArg*> input_args;
+  for (const auto& input_name : split_graphs_info_.user_input_names) {
+    input_args.emplace_back(graph.GetNodeArg(input_name));
+  }
+
+  // Remove the training initializers from the graph and move them to graph inputs.
+  for (const auto& initializer_name : split_graphs_info_.initializer_names_to_train) {
+    input_args.emplace_back(graph.GetNodeArg(initializer_name));
+    graph.RemoveInitializedTensor(initializer_name);
+  }
+
+  graph.SetInputs(input_args);
+  return Status::OK();
+}
+
+// Build the gradient graph and split it to forward and backward graphs.
+// Since the input shapes may differ, and the graph optimizers (mainly constant folding) may fold this
+// shape info to constants, so the optimized graph (before gradient graph building) can not be shared.
+// So each time we need to start from the beginning, i.e., 1) replace input shapes, 2) apply graph optimizers,
```

Comment:
Are we invoking to rebuild the gradient graph for every batch? 
If yes, we need to come up with a better way to handle this. It might not be too back to just disabled the shape-related constant folding.

Reply:
I see how this is used now... if the new input shape is different from the current shape, we rebuild the gradient graph. 

This is not a sustainable solution. I am adding an item to track this.